numpy/_array_api/_array_object.py

```diff
@@ -0,0 +1,662 @@
+        """
+        Performs the operation __setitem__.
+        """
+        # Note: Only indices required by the spec are allowed. See the
+        # docstring of _validate_index
+        key = self._validate_index(key, self.shape)
+        res = self._array.__setitem__(key, asarray(value)._array)
+        return self.__class__._new(res)
+
+    def __sub__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __sub__.
+        """
+        res = self._array.__sub__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __truediv__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __truediv__.
+        """
+        res = self._array.__truediv__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __xor__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __xor__.
+        """
+        res = self._array.__xor__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __iadd__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __iadd__.
+        """
+        res = self._array.__iadd__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __radd__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __radd__.
+        """
+        res = self._array.__radd__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __iand__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __iand__.
+        """
+        res = self._array.__iand__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __rand__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __rand__.
+        """
+        res = self._array.__rand__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __ifloordiv__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __ifloordiv__.
+        """
+        res = self._array.__ifloordiv__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __rfloordiv__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __rfloordiv__.
+        """
+        res = self._array.__rfloordiv__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __ilshift__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __ilshift__.
+        """
+        res = self._array.__ilshift__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __rlshift__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __rlshift__.
+        """
+        res = self._array.__rlshift__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __imatmul__(self: array, other: array, /) -> array:
```

Comment:
Maybe because matmul can't truly be done in-place, since it produces an array of a different shape, so it's unnecessary to implement it (by default `a OP= b` just does `a = a OP b` if `a.__iOP__` isn't implemented). That probably means this function here doesn't actually work, though, so we should remedy that.